src/main/java/com/maxdemarzi/shortest/Service.java

```diff
@@ -438,6 +415,96 @@ private void streamShortestPathsUsingHandwrittenBFS(String centerEmail, List<Str
         }
     }
 
+    private static volatile IntIntMap relationships;
+
+    private static final IntIntMap relationshipCosts(ReadOperations readOps) {
+        if (relationships == null) {
+            Map<String,Integer> costs = ImmutableMap.<String, Integer>builder()
+                .put("EqualTo", 1)
+                .put("HasEmail", 1)
+                .put("AuthoredBy", 2)
+                .put("ContainsEmail", 2)
+                .put("CoAuthorOf", 4)
+                .put("Follows", 4)
+                .put("hasContact", 4)
+                .put("HasUrl", 4)
+                .build();
+            Builder<Integer, Integer> builder = ImmutableMap.<Integer, Integer>builder();
+            for (Map.Entry<String, Integer> e : costs.entrySet()) {
+                builder.put(readOps.relationshipTypeGetForName(e.getKey()), e.getValue().intValue());
+            }
+            relationships = HashIntIntMaps.getDefaultFactory()
+                .withDefaultValue(100) // makes it too expensive to explore relationships not listed here
+                .newImmutableMap(builder.build());
+        }
+        return relationships;
+    }
+
+    private void streamShortestPathsUsingDijkstra(String centerEmail, List<String> bibEntries, List<String> edgeEmails, int maxLength, JsonGenerator jg) {
+        try (Transaction tx = db.beginTx()) {
+            final Long centerNodeId;
+            try {
+                centerNodeId = nodeCache.getEmailNode(centerEmail);
+            } catch (ExecutionException e) {
+                return;
+            }
+            Map<Long, Integer> startNodes = HashLongIntMaps.newMutableMap();
+            for (Long nodeId : nodeCache.getBibliographEntryNodes(bibEntries))  {
+                startNodes.put(nodeId, 1);
+            }
+            startNodes.put(centerNodeId, 0);
+
+            final HashLongObjMap<String> edgeEmailsByNodeId = HashLongObjMaps.newMutableMap();
+            for (String edgeEmail : edgeEmails) {
+                try {
+                    edgeEmailsByNodeId.put(nodeCache.getEmailNode(edgeEmail), edgeEmail);
+                } catch (Exception e) {
+                    continue;
+                }
+            }
+
+            ThreadToStatementContextBridge ctx = dbAPI.getDependencyResolver().resolveDependency(ThreadToStatementContextBridge.class);
+            ReadOperations ops = ctx.get().readOperations();
+
+            Dijkstra dijkstra = new Dijkstra(ops, relationshipCosts(ops), startNodes, maxLength, new Traversal.NodeCallback() {
```

Comment:
do we need to multiply maxLength by 4 at some point because it's being treated as max cost?